src/bit-systems/audio-zone-system.ts

```diff
@@ -0,0 +1,318 @@
+      .filter(zone => AudioZone.flags[zone] & AUDIO_ZONE_FLAGS.OUT_IN)
+      .filter(exclude(sourceZones))
+      .filter(hasIntersection(ray))
+      .map(zone => getZoneAudioParams(zone))
+      .reduce(paramsReducer, null);
+
+    if (!outInParams && !inOutParams) {
+      restoreSourceParams(source);
+    } else if (outInParams && !inOutParams) {
+      applySourceParams(source, outInParams);
+    } else if (!outInParams && inOutParams) {
+      applySourceParams(source, inOutParams);
+    } else {
+      // In this case two zones ar acting over the same source simultaneously.
+      // We apply the closest zone params with the lowest gain
+      applySourceParams(
+        source,
+        Object.assign(
+          {},
+          inOutParams,
+          outInParams,
+          paramsReducer(
+            {
+              gain: outInParams?.gain
+            },
+            {
+              gain: inOutParams?.gain
+            }
+          )
+        )
+      );
+    }
+  };
+})();
+
+const REDUCED_KEYS = [
+  "gain",
+  "maxDistance",
+  "refDistance",
+  "rolloffFactor",
+  "coneInnerAngle",
+  "coneOuterAngle",
+  "coneOuterGain"
+] as const;
+type ReducedAudioSettingsKeys = typeof REDUCED_KEYS[number];
+type ReducedAudioSettings = Pick<AudioSettings, ReducedAudioSettingsKeys>;
+
+// We apply the most restrictive audio parameters
+const paramsReducer = (acc: AudioSettings, curr: AudioSettings): AudioSettings => {
+  if (!curr && !acc) return {} as AudioSettings;
+  else if (curr && !acc) return curr;
+  else if (!curr && acc) return acc;
+  else
+    return REDUCED_KEYS.reduce((result: ReducedAudioSettings, key: ReducedAudioSettingsKeys): AudioSettings => {
+      if (curr[key] !== undefined && acc[key] !== undefined) {
+        result[key] = Math.min(acc[key]!, curr[key]!);
+      } else if (curr[key] !== undefined && acc[key] === undefined) {
+        result[key] = curr[key];
+      } else if (curr[key] === undefined && acc[key] !== undefined) {
+        result[key] = acc[key];
+      }
+      return result;
+    }, {});
+};
+
+const addOrRemoveZone = (zones: Set<number>, zone: number, position: Vector3) => {
+  if (!zones) return;
+  const isInZone = isZoneEnabled(zone) && zoneContains(zone, position);
+  const wasInZone = zones.has(zone);
+  if (isInZone && !wasInZone) {
+    zones.add(zone);
+  } else if (!isInZone && wasInZone) {
+    zones.delete(zone);
+  }
+};
+
+const currZones = new Map<number, Set<number>>();
+const prevZones = new Map<number, Set<number>>();
+const aabbs = new Map<number, Box3>();
+
+const listenerPos = new Vector3();
+const audioZoneQuery = defineQuery([AudioZone]);
+const audioZoneEnterQuery = enterQuery(audioZoneQuery);
+const audioZoneExitQuery = exitQuery(audioZoneQuery);
+const audioZoneSourceQuery = defineQuery([AudioEmitter]);
+const audioZoneSourceEnterQuery = enterQuery(audioZoneSourceQuery);
+const audioZoneSourceExitQuery = exitQuery(audioZoneSourceQuery);
```

Comment:
The query looks for `AudioEmitter`, so these names might be clearer:

 `audioEmitterQuery`, `audioEmitterEnterQuery`, `audioEmitterExitQuery` OR
 `audioEmittersQuery`, `enteredAudioEmittersQuery`, `exitedAudioEmittersQuery`.
 
 (We should probably pick one or the other pattern and use it everywhere, but we haven't been consistent between these two.)

Reply:
Makes sense to consolidate to emitter everywhere so I'm going to change all strings to refect that. And I agree on the styling guide. We might benefit from having a more structured naming convention for things like:
- Component names (when to include "Tag" or other Pre/Suffixes)
- Queries names (enter[Component]Query, [Component]enterQuery, enter[Component], etc)
- Eid naming in queries, loops, variables, etc. No we use just eid everywhere, I tried using [component]Eid as much as possible to avoid confusion. Maybe other naming conventions.
- Inflators naming seem to be more consistent.